pallets/phala/src/stakepool.rs

```diff
@@ -759,85 +742,38 @@ pub mod pallet {
 			Ok(())
 		}
 
-		/// Claims staker's pending rewards of the sender and send to the `target`
-		///
-		/// The rewards associate to sender's "owner role" will not be claimed
-		///
-		/// Requires:
-		///
-		/// 1. The sender is a staker
-		#[pallet::weight(0)]
-		pub fn claim_staker_rewards(
-			origin: OriginFor<T>,
-			pid: u64,
-			target: T::AccountId,
-		) -> DispatchResult {
-			let who = ensure_signed(origin)?;
-			let pool_info = Self::ensure_pool(pid)?;
-			let info_key = (pid, who.clone());
-			let mut user_info = Self::pool_stakers(&info_key).ok_or(Error::<T>::NoRewardToClaim)?;
-			pool_info.settle_user_pending_reward(&mut user_info);
-			let rewards = user_info.available_rewards;
-			ensure!(rewards > Zero::zero(), Error::<T>::NoRewardToClaim);
-			mining::Pallet::<T>::withdraw_subsidy_pool(&target, rewards)
-				.or(Err(Error::<T>::InternalSubsidyPoolCannotWithdraw))?;
-			user_info.available_rewards = Zero::zero();
-			// Update ledger
-			StakePools::<T>::insert(pid, &pool_info);
-			PoolStakers::<T>::insert(&info_key, &user_info);
-			Self::deposit_event(Event::<T>::StakerRewardsWithdrawn {
-				pid,
-				user: who,
-				amount: rewards,
-			});
-
-			Ok(())
-		}
-
-		/// Claims all the pending rewards of the sender and send to the `target`
-		///
-		/// Requires:
+		/// Let any user to launch a stakepool withdraw. Then check if the pool need to be forced shutdown.
 		///
-		/// 1. The sender is a pool owner or staker
+		/// If the shutdown condition is met, all workers in the pool will be forced shutdown.
+		/// Note: This function doesn't guarantee no-op when there's error.
 		#[pallet::weight(0)]
-		pub fn claim_rewards(
-			origin: OriginFor<T>,
-			pid: u64,
-			target: T::AccountId,
-		) -> DispatchResult {
-			let who = ensure_signed(origin)?;
-			let mut pool_info = Self::ensure_pool(pid)?;
-			let mut rewards = BalanceOf::<T>::zero();
-			// Add pool owner's reward if applicable
-			if who == pool_info.owner {
-				rewards += pool_info.owner_reward;
-				pool_info.owner_reward = Zero::zero();
-			}
-			// Settle the pending reward, and calculate the rewards belong to user
-			let info_key = (pid, who.clone());
-			let mut user_info = Self::pool_stakers(&info_key);
-			if let Some(ref mut user_info) = user_info {
-				pool_info.settle_user_pending_reward(user_info);
-				rewards += user_info.available_rewards;
-				user_info.available_rewards = Zero::zero();
+		pub fn check_and_maybe_force_withdraw(origin: OriginFor<T>, pid: u64) -> DispatchResult {
+			ensure_signed(origin)?;
+			let now = <T as registry::Config>::UnixTime::now()
+				.as_secs()
+				.saturated_into::<u64>();
+			let mut pool = Self::ensure_pool(pid)?;
+			Self::try_process_withdraw_queue(&mut pool);
```

Comment:
`pool` was changed but never saved to the storage

Reply:
done, insert poolinfo before calling do_stop_mining and after processing withdraw queue